### ISSUE TITLE
Found a problem with stored procedure interface

### ISSUE THREAD

**Issue by CycoPH** [TimelordUK/node-sqlserver-v8#58]
Problem description:
When calling the same stored proc in very short succession the input parameters to the stored proc get mixed up. Short example.  I have a [getBalance] stored proc and when calling it 5 times within XXms then only the first invocation returns the correct data, the others all return the data from the last call.

Solution search:
I've started looking at the implementation of the stored proc interface in lib/procedure.js.
Starting with "callproc()" we follow the call to createProcedure().  Here the first "problem" occurs.
If you call the SP very quickly the "cache[name]" will be undefined until the "describeProcedure()" function actually puts a valid "BoundProcedure" into the cache. So you might get a couple of BoundProcedure objects being clobbered.

The problem comes in with the "meta" data used in the BoundProcedure. The meta.params structure is changed by the bindParams() function everytime you call a SP.

Problem:
1. Call SP getBalance("user1").  A ".val" field is being set in the meta.params object. The SP is executed.
2. Now call this SP a couple more times
  getBalance("user2")
  getBalance("user3")
Every time the same meta.params (which is cached) object is changed.
3. When the driver finally gets time to actually do the calling you effectively call the same stored proc with "user3" as input.

Solution:
I just fixed this as a test in the privateCall function:
    function privateCall (notify, params, cb) {
      var myMeta = JSON.parse(JSON.stringify(meta));
      bindParams(myMeta, params)
      if (cb) {
        callStoredProcedure(notify, myMeta.signature, myMeta.params, function (err, results, output) {
          cb(err, results, output)
        })
      } else {
        callStoredProcedure(notify, myMeta.signature, myMeta.params)
      }
    }
Basically do a deep-clone of the meta object and only bind the parameters to that clone, and execute the SP with the clone.
I did not look at performance, or anything else,  just wanted to fix the problem so that my production system can continue running.  I'm sure there are better clone options than JSON.parse :)

**Comment by TimelordUK** [comment]
thanks for the detailed explanation and bringing this to my attention - I am working on a fix now.

**Comment by TimelordUK** [comment]
ok thanks this should be fixed in most recent release.

**Comment by TimelordUK** [comment]
thanks for raising this, I have added unit tests for this, indeed there was a couple of issues leading to this problem I believe I now have got to bottom of it, so closing
